didasm/src/tokens.rs

```diff
@@ -0,0 +1,352 @@
+        b: Register,
+        x: Register
+    },
+    /// [<REG>][<REG>+]
+    /// [<REG>+<REG>+]
+    RegSumAutoincrement {
+        b: Register,
+        x: Register
+    },
+    /// [<REG>][<REG>-]
+    /// [<REG>+<REG>-]
+    RegSumAutodecrement {
+        b: Register,
+        // x: Register
+    },
+    /// [<REG>]+<EXPR>
+    /// <EXPR>[<REG>]
+    /// [<REG>+<EXPR>]
+    /// [<REG>].<EXPR>
+    Based {
+        b: Register,
+        depls: Expr
+    },
+    /// [<REG>]+<EXPR>
+    /// <EXPR>[<REG>]
+    /// [<REG>+<EXPR>]
+    /// [<REG>].<EXPR>
+    Indexed {
+        x: Register,
+        depls: Expr
+    },
+    /// [<REG>][<REG>]+<EXPR>
+    /// <EXPR>[<REG>][<REG>]
+    /// [<REG>+<REG>+<EXPR>]
+    /// [<REG>][<REG>].<EXPR>
+    BasedIndexed {
+        b: Register,
+        x: Register,
+        depls: Expr
+    }
+}
+
+use RegOrExpr::*;
+enum RegOrExpr {
+    Reg(Register),
+    Expression(Expr)
+}
+
+impl FromStr for RegOrExpr {
+    type Err = String;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Ok(reg) = Register::from_str(s) {
+            Ok(RegOrExpr::Reg(reg))
+        } else {
+            Ok(RegOrExpr::Expression(Expr::from_str(s)?))
+        }
+    }
+}
+
+impl FromStr for Operand {
+    type Err = String;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        static CHECK_IF_DOUBLE_INDIRECT: Lazy<Regex> = Lazy::new(|| {
+            Regex::new(r"^\[\s*\[\s*(.+?)\s*\]\s*\]$").unwrap()
+        });
+        static CHECK_OUTER_BRACKETS: Lazy<Regex> = Lazy::new(|| {
+            Regex::new(r"^\[\s*(.+)\s*\]$").unwrap()
+        });
+        static CHECK_INSIDE_BRACKETS: Lazy<Regex> = Lazy::new(|| {
+            Regex::new(r"^([a-z0-9_]+)\s*\+\s*([a-z0-9_]+)\s*(?:\+\s*([a-z0-9_]+)|([+\-]))?$").unwrap()
+        });
+
+
+
+        if let Some(cap) = CHECK_IF_DOUBLE_INDIRECT.captures(s) {
+            let (_, [op]) = cap.extract();
+            if let Ok(_) = Register::from_str(op) {
+                Err(format!("This indirect addressing syntax cannot be done on registers! ({op})"))
+            } else {
+                match Expr::from_str(op) {
+                    Ok(s) => Ok(Operand::Indirect(s)),
+                    Err(e) => Err(format!("Syntax error on indirect addressing '{op}': {e}"))
+                }
+            }
+        } else if let Some(cap) = CHECK_OUTER_BRACKETS.captures(s) {
+            let (_, [op]) = cap.extract();
+            if let Ok(reg) = Register::from_str(op) {
+                if matches!(reg, Ra|Rb|Rc|Is) {
+                    Err(format!("{reg} is not supported in indirect register addresing!"))
+                } else {
+                    Ok(Operand::RegIndirect(reg))
+                }
+            } else if let Ok(expr) = Expr::from_str(op) {
+                Ok(Operand::Direct(expr))
+            } else if let Some(cap) = CHECK_INSIDE_BRACKETS.captures(op) {
+                let op1 = RegOrExpr::from_str(cap.get(1).unwrap().as_str())?;
```

Comment:
don't use unwrap

Reply:
If the expression fails here, the error is internal and unrecoverable. I can use an expect to signal it is an internal error but it adds too much error checking for something that might not be worth doing it for